File: ios/Classes/NavigationLauncherPlugin.swift
import Flutter
import UIKit
import MapKit

public class NavigationLauncherPlugin: NSObject, FlutterPlugin {
    public static func register(with registrar: FlutterPluginRegistrar) {
        let channel = FlutterMethodChannel(name: "navigation_launcher", binaryMessenger: registrar.messenger())
        let instance = NavigationLauncherPlugin()
        registrar.addMethodCallDelegate(instance, channel: channel)
    }
    
    public func handle(_ call: FlutterMethodCall, result: @escaping FlutterResult) {
        switch call.method {
        case "getInstalledMaps":
            var list = [String]()
            for app in MapApp.allCases {
                if isInstalled(url: app.url) {
                    list.append(app.value)
                }
            }
            result(list)
        case "launchNavigation":
            let args = call.arguments as! [String: Any]
            let app = MapApp.allCases.first(where: {a in a.value == args["app"] as! String})
            if(app == nil){
                result(FlutterError(code: "unknown_app", message: "不支持的地图", details: nil))
            } else if !isInstalled(url: app!.url) {
                result(FlutterError(code: app!.value, message: "未安装\(app!.desc)地图", details: nil))
            }else{
                let point = args["point"] as! [Double]
                let name = args["name"] as! String
                switch app!{
                case MapApp.gaode:
                    navigateGaode(latitude: point[1], longitude: point[0], destinationName: name)
                case MapApp.baidu:
                    navigateBaidu(latitude: point[1], longitude: point[0], destinationName: name)
                case MapApp.tencent:
                    navigateTencent(latitude: point[1], longitude: point[0], destinationName: name)
                case MapApp.google:
                    navigateGoogle(latitude: point[1], longitude: point[0], destinationName: name)
                case MapApp.apple:
                    navigateApple(latitude: point[1], longitude: point[0], destinationName: name)
                }
            }
        default:
            result(FlutterMethodNotImplemented)
        }
    }
    
    func isInstalled(url: String) -> Bool{
        return UIApplication.shared.canOpenURL(URL(string: url)!)
    }
    
    func navigateApple(latitude: Double, longitude: Double, destinationName: String) {
        let coordinate = CLLocationCoordinate2D(latitude: latitude, longitude: longitude)
        let destinationPlacemark = MKPlacemark(coordinate: coordinate, addressDictionary: nil)
        let destinationMapItem = MKMapItem(placemark: destinationPlacemark)
        destinationMapItem.name = destinationName
        let launchOptions = [MKLaunchOptionsDirectionsModeKey: MKLaunchOptionsDirectionsModeDriving]
        MKMapItem.openMaps(with: [destinationMapItem], launchOptions: launchOptions)
    }
    
    func navigateGoogle(latitude: Double, longitude: Double, destinationName: String) {
        let urlString = "comgooglemaps://?saddr=&daddr=\(latitude),\(longitude)&directionsmode=driving"
        UIApplication.shared.open(URL(string: urlString)!, options: [:], completionHandler: nil)
    }
    
    func navigateGaode(latitude: Double, longitude: Double, destinationName: String) {
        let urlString = "iosamap://path?dlat=\(latitude)&dlon=\(longitude)&dname=\(destinationName)&dev=1&t=0"
        UIApplication.shared.open(URL(string: urlString)!, options: [:], completionHandler: nil)
    }
    
    func navigateBaidu(latitude: Double, longitude: Double, destinationName: String) {
        let urlString = "baidumap://map/direction?destination=\(latitude),\(longitude)&coord_type=wgs84&mode=driving&src=\(String(describing: Bundle.main.bundleIdentifier))"
        UIApplication.shared.open(URL(string: urlString)!, options: [:], completionHandler: nil)
    }
    
    func navigateTencent(latitude: Double, longitude: Double, destinationName: String) {
        let urlString = "qqmap://map/routeplan?type=drive&fromcoord=CurrentLocation&to=\(destinationName)&tocoord=\(latitude),\(longitude)"
        UIApplication.shared.open(URL(string: urlString)!, options: [:], completionHandler: nil)
    }
}

enum MapApp :CaseIterable{
    case gaode
    case baidu
    case tencent
    case google
    case apple

    var value: String {
        switch self {
        case .gaode:
            return "gaode"
        case .baidu:
            return "baidu"
        case .tencent:
            return "tencent"
        case .google:
            return "google"
        case .apple:
            return "apple"
        }
    }

    var url: String {
        switch self {
        case .gaode:
            return "iosamap://"
        case .baidu:
            return "baidumap://"
        case .tencent:
            return "qqmap://"
        case .google:
            return "comgooglemaps://"
        case .apple:
            return "maps://"
        }
    }

    var desc: String {
        switch self {
        case .gaode:
            return "高德"
        case .baidu:
            return "百度"
        case .tencent:
            return "腾讯"
        case .google:
            return "谷歌"
        case .apple:
            return "苹果"
        }
    }
}

File: android/src/main/kotlin/com/whale/navigation_launcher/NavigationLauncherPlugin.kt
package com.whale.navigation_launcher

import android.content.Context
import android.content.Intent
import android.net.Uri
import io.flutter.embedding.engine.plugins.FlutterPlugin
import io.flutter.plugin.common.MethodCall
import io.flutter.plugin.common.MethodChannel
import io.flutter.plugin.common.MethodChannel.MethodCallHandler
import io.flutter.plugin.common.MethodChannel.Result

class NavigationLauncherPlugin : FlutterPlugin, MethodCallHandler {
    private lateinit var channel: MethodChannel
    private lateinit var context: Context

    override fun onAttachedToEngine(flutterPluginBinding: FlutterPlugin.FlutterPluginBinding) {
        channel = MethodChannel(flutterPluginBinding.binaryMessenger, "navigation_launcher")
        channel.setMethodCallHandler(this)
        context = flutterPluginBinding.applicationContext
    }

    override fun onMethodCall(call: MethodCall, result: Result) {
        if (call.method == "getInstalledMaps") {
            val list = arrayListOf<String>()
            for (value in MapApp.values()) {
                if (isInstalled(value.id)) list.add(value.value)
            }
            result.success(list)
        } else if (call.method == "launchNavigation") {
             val app = MapApp.values().firstOrNull {
                it.value == call.argument<String>("app")
            }
            if (app == null) {
                result.error("unknown_app", "不支持的地图", null)
            } else if (!isInstalled(app.id)) {
                result.error(app.value, "未安装${app.desc}地图", null)
            } else {
                val point = call.argument<List<Double>>("point")
                val name = call.argument<String>("name")
                when (app) {
                    MapApp.GAODE -> navigateGaode(point!![0], point[1], name)
                    MapApp.BAIDU -> navigateBaidu(point!![0], point[1])
                    MapApp.TENCENT -> navigateTencent(point!![0], point[1], name)
                    MapApp.GOOGLE -> navigateGoogle(point!![0], point[1])
                }
                result.success(true)
            }
        } else {
            result.notImplemented()
        }
    }

    private fun navigateGaode(lon: Double, lat: Double, name: String?) {
        Intent().apply {
            action = Intent.ACTION_VIEW
            val param = "dlat=${lat}&dlon=${lon}&dname=${name}&dev=1&t=0"
            data = Uri.parse("amapuri://route/plan/?$param")
            addFlags(Intent.FLAG_ACTIVITY_NEW_TASK)
            context.startActivity(this)
        }
    }

    private fun navigateBaidu(lon: Double, lat: Double) {
        Intent().apply {
            action = Intent.ACTION_VIEW
            val param =
                "destination=${lat},${lon}&coord_type=wgs84&mode=driving&src=${context.packageName}"
            data = Uri.parse("baidumap://map/direction?$param")
            addFlags(Intent.FLAG_ACTIVITY_NEW_TASK)
            context.startActivity(this)
        }
    }

    private fun navigateTencent(lon: Double, lat: Double, name: String?) {
        Intent().apply {
            action = Intent.ACTION_VIEW
            val param = "type=drive&fromcoord=CurrentLocation&to=${name}&tocoord=${lat},${lon}"
            data = Uri.parse("qqmap://map/routeplan?$param")
            addFlags(Intent.FLAG_ACTIVITY_NEW_TASK)
            context.startActivity(this)
        }
    }

    private fun navigateGoogle(lon: Double, lat: Double) {
        Intent().apply {
            action = Intent.ACTION_VIEW
            val param = "q=${lat},${lon}"
            data = Uri.parse("google.navigation:$param")
            addFlags(Intent.FLAG_ACTIVITY_NEW_TASK)
            setPackage(MapApp.GOOGLE.id)
            context.startActivity(this)
        }
    }

    @Suppress("DEPRECATION")
    private fun isInstalled(id: String): Boolean {
        val packages = context.packageManager.getInstalledPackages(0)
        for (p in packages) {
            if (p.packageName.equals(id)) return true else continue
        }
        return false
    }

    override fun onDetachedFromEngine(binding: FlutterPlugin.FlutterPluginBinding) {
        channel.setMethodCallHandler(null)
    }

    enum class MapApp(val value: String, val id: String, val desc: String) {
        GAODE("gaode", "com.autonavi.minimap", "高德"),
        BAIDU("baidu", "com.baidu.BaiduMap", "百度"),
        TENCENT("tencent", "com.tencent.map", "腾讯"),
        GOOGLE("google", "com.google.android.apps.maps", "谷歌");
    }
}
